File: AssetSync/src/AssetSync.Infrastructure/Api/IntuneService.cs
using System.Text.RegularExpressions;
using AssetSync.Core.Interfaces;
using AssetSync.Core.Services;
using Device = AssetSync.Core.Models.Device;
using Azure.Identity;
using Microsoft.Graph.Models;
using Microsoft.Kiota.Authentication.Azure;
using Microsoft.Kiota.Http.HttpClientLibrary;

namespace AssetSync.Infrastructure.Api;

public class IntuneService : IIntuneService
{
    private const int PageSize = 1000;
    // Matches PM followed by exactly 7 digits anywhere in a string (e.g. Notes field)
    private static readonly Regex PmTagRegex = new(@"\bPM\d{7}\b", RegexOptions.IgnoreCase);
    private readonly string _tenantId;
    private readonly string _clientId;
    private readonly string _clientSecret;
    private readonly Func<string> _getClientSecret;

    public IntuneService(string tenantId, string clientId, Func<string> getClientSecret)
    {
        _tenantId = tenantId;
        _clientId = clientId;
        _getClientSecret = getClientSecret;
        _clientSecret = string.Empty;
    }

    public IntuneService(string tenantId, string clientId, string clientSecret)
    {
        _tenantId = tenantId;
        _clientId = clientId;
        _clientSecret = clientSecret;
        _getClientSecret = () => _clientSecret;
    }

    public async Task<IReadOnlyList<Device>> GetManagedDevicesAsync(CancellationToken cancellationToken = default)
    {
        var credential = new ClientSecretCredential(_tenantId, _clientId, _getClientSecret());
        var authProvider = new AzureIdentityAuthenticationProvider(credential, scopes: new[] { "https://graph.microsoft.com/.default" });
        var adapter = new HttpClientRequestAdapter(authProvider);
        var client = new Microsoft.Graph.GraphServiceClient(adapter);

        var list = new List<Device>();
        var page = await client.DeviceManagement.ManagedDevices.GetAsync(c => c.QueryParameters.Top = PageSize, cancellationToken).ConfigureAwait(false);
        await CollectPageAsync(page, list).ConfigureAwait(false);

        while (page?.OdataNextLink != null)
        {
            page = await client.DeviceManagement.ManagedDevices.WithUrl(page.OdataNextLink!).GetAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
            await CollectPageAsync(page, list).ConfigureAwait(false);
        }

        return list;
    }

    public async Task<bool> WriteBackAssetTagAsync(string intuneDeviceId, string assetTag, string? existingNotes, CancellationToken cancellationToken = default)
    {
        try
        {
            var credential = new ClientSecretCredential(_tenantId, _clientId, _getClientSecret());
            var authProvider = new AzureIdentityAuthenticationProvider(credential, scopes: new[] { "https://graph.microsoft.com/.default" });
            var adapter = new HttpClientRequestAdapter(authProvider);
            var client = new Microsoft.Graph.GraphServiceClient(adapter);

            // Build updated notes: replace existing PM tag line if present, otherwise append
            string updatedNotes;
            var tagLine = $"Asset Tag: {assetTag}";
            if (!string.IsNullOrEmpty(existingNotes))
            {
                // Replace any existing "Asset Tag: PMxxxxxxx" line, or append if not found
                var replaced = System.Text.RegularExpressions.Regex.Replace(
                    existingNotes,
                    @"Asset Tag:\s*PM\d{7}",
                    tagLine,
                    System.Text.RegularExpressions.RegexOptions.IgnoreCase);
                updatedNotes = replaced == existingNotes
                    ? existingNotes.TrimEnd() + "\n" + tagLine
                    : replaced;
            }
            else
            {
                updatedNotes = tagLine;
            }

            // Uses the Intune managed device ID (m.Id), NOT the Azure AD device ID
            var body = new ManagedDevice { Notes = updatedNotes };
            await client.DeviceManagement.ManagedDevices[intuneDeviceId].PatchAsync(body, cancellationToken: cancellationToken).ConfigureAwait(false);
            return true;
        }
        catch (Exception ex)
        {
            // Re-throw so SyncEngine can log the real error detail
            throw new InvalidOperationException($"Intune write-back failed for device {intuneDeviceId}: {ex.Message}", ex);
        }
    }

    private static Task CollectPageAsync(ManagedDeviceCollectionResponse? page, List<Device> list)
    {
        if (page?.Value == null) return Task.CompletedTask;
        foreach (var m in page.Value)
        {
            var serial = m.SerialNumber ?? "";
            // Extract a PM-format asset tag from the Notes field if present
            string? mdmAssetTag = null;
            if (!string.IsNullOrEmpty(m.Notes))
            {
                var match = PmTagRegex.Match(m.Notes);
                if (match.Success) mdmAssetTag = match.Value.ToUpperInvariant();
            }
            list.Add(new Device
            {
                NormalizedSerial = SerialNumberNormalizer.Normalize(serial),
                SerialNumber = serial,
                DeviceName = m.DeviceName,
                Model = m.Model,
                AssignedUserUpn = m.UserPrincipalName,
                OsVersion = m.OsVersion,
                DeviceType = m.DeviceCategoryDisplayName ?? m.ManagementAgent?.ToString(),
                PlatformSource = "Intune",
                AzureAdDeviceId = m.AzureADDeviceId,
                OperatingSystem = m.OperatingSystem,
                MdmAssetTag = mdmAssetTag,
                IntuneNotes = m.Notes,
                IntuneDeviceId = m.Id
            });
        }
        return Task.CompletedTask;
    }
}

File: AssetSync-macOS/AssetSyncInfrastructure/Sources/AssetSyncInfrastructure/Api/IntuneService.swift
import Foundation
import MSAL
import AssetSyncCore

/// Fetches managed devices from Microsoft Graph via MSAL client credentials flow.
public final class IntuneService: IntuneServiceProtocol, @unchecked Sendable {
    private static let pageSize = 1000
    private static let graphBase = "https://graph.microsoft.com/v1.0"
    private static let pmTagRegex = /\bPM\d{7}\b/
    private let tenantId: String
    private let clientId: String
    private let getClientSecret: @Sendable () -> String
    private let session: URLSession

    public init(tenantId: String, clientId: String, getClientSecret: @escaping @Sendable () -> String,
                session: URLSession = .shared) {
        self.tenantId = tenantId
        self.clientId = clientId
        self.getClientSecret = getClientSecret
        self.session = session
    }

    public func getManagedDevices() async throws -> [Device] {
        let token = try await acquireToken()
        var devices: [Device] = []
        var url: String? = "\(Self.graphBase)/deviceManagement/managedDevices?$top=\(Self.pageSize)"

        while let currentURL = url {
            var request = URLRequest(url: URL(string: currentURL)!)
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            let (data, _) = try await session.data(for: request)
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            guard let values = json?["value"] as? [[String: Any]] else { break }

            for m in values {
                let serial = m["serialNumber"] as? String ?? ""
                let notes = m["notes"] as? String
                var mdmAssetTag: String?
                if let notes, let match = notes.firstMatch(of: Self.pmTagRegex) {
                    mdmAssetTag = String(match.output).uppercased()
                }

                var device = Device()
                device.normalizedSerial = SerialNumberNormalizer.normalize(serial)
                device.serialNumber = serial
                device.deviceName = m["deviceName"] as? String
                device.model = m["model"] as? String
                device.assignedUserUpn = m["userPrincipalName"] as? String
                device.osVersion = m["osVersion"] as? String
                device.deviceType = (m["deviceCategoryDisplayName"] as? String) ?? (m["managementAgent"] as? String)
                device.platformSource = "Intune"
                device.azureAdDeviceId = m["azureADDeviceId"] as? String
                device.operatingSystem = m["operatingSystem"] as? String
                device.mdmAssetTag = mdmAssetTag
                device.intuneNotes = notes
                device.intuneDeviceId = m["id"] as? String
                devices.append(device)
            }

            url = json?["@odata.nextLink"] as? String
        }

        return devices
    }

    public func writeBackAssetTag(intuneDeviceId: String, assetTag: String, existingNotes: String?) async throws -> Bool {
        let token = try await acquireToken()
        let tagLine = "Asset Tag: \(assetTag)"
        let updatedNotes: String
        if let existing = existingNotes, !existing.isEmpty {
            let replaced = existing.replacingOccurrences(
                of: #"Asset Tag:\s*PM\d{7}"#,
                with: tagLine,
                options: .regularExpression)
            updatedNotes = (replaced == existing)
                ? existing.trimmingCharacters(in: .whitespacesAndNewlines) + "\n" + tagLine
                : replaced
        } else {
            updatedNotes = tagLine
        }

        var request = URLRequest(url: URL(string: "\(Self.graphBase)/deviceManagement/managedDevices/\(intuneDeviceId)")!)
        request.httpMethod = "PATCH"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["notes": updatedNotes])

        let (_, response) = try await session.data(for: request)
        return (response as? HTTPURLResponse)?.statusCode ?? 0 < 300
    }

    // MARK: - MSAL Token Acquisition

    private func acquireToken() async throws -> String {
        let authority = try MSALAADAuthority(url: URL(string: "https://login.microsoftonline.com/\(tenantId)")!)
        let config = MSALPublicClientApplicationConfig(clientId: clientId, redirectUri: nil, authority: authority)
        // For confidential client, use MSAL's token acquisition with client credential
        // Note: MSAL for macOS uses MSALPublicClientApplication; for daemon/service flows
        // you'll use a direct OAuth2 client_credentials POST instead:
        return try await acquireTokenViaClientCredentials()
    }

    private func acquireTokenViaClientCredentials() async throws -> String {
        let url = URL(string: "https://login.microsoftonline.com/\(tenantId)/oauth2/v2.0/token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let body = [
            "client_id": clientId,
            "client_secret": getClientSecret(),
            "scope": "https://graph.microsoft.com/.default",
            "grant_type": "client_credentials",
        ].map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.value)" }
            .joined(separator: "&")
        request.httpBody = Data(body.utf8)

        let (data, _) = try await session.data(for: request)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["access_token"] as? String else {
            throw AssetSyncError.snipeItRejected("Failed to acquire Intune token")
        }
        return token
    }
}
